app/web_framework/cluster_application.py

```diff
@@ -15,6 +52,33 @@ def get_all_handlers(root_route, default_params):
         :rtype: list [tuple (str, tornado.web.RequestHandler, dict)]
         """
         all_route_nodes = [root_route] + root_route.descendants()
+
+        # Map regex to all the routes that have the same regex pattern.
+        route_map = {}
+        for route in all_route_nodes:
+            route_path = route.regex()
+            if route_path not in route_map:
+                route_map[route_path] = []
+            route_map[route_path].append(route)
+
+        # Map regex to a list of route and RequestHandlerWrapper tuples.
+        # The RequestHandlerWrapper knows all the RequstHandlers and what API version those handlers work
+        # with for a given route, so when a route is hit by the API, we know which RequestHandler to use.
+        route_handler_map = {}
+        for regex in route_map:
+            for route in route_map[regex]:
+                if regex not in route_handler_map:
+                    route_handler_map[regex] = []
+                # Create a dict that maps version to handler
+                handlers_map = dict([(route.version, route.handler) for route in route_map[regex]])
```

Comment:
Ah yeah nice catch, we can move this outside of this inner loop to make things a little more efficient. 

So it's a little tricky what's going on here (and if you see any areas for improvement let me know), but what's happening is we're taking our map of regex to list of routes that share the same regex, and we want to create a map of regex to list of route & RequestHandlerWrapper tuples.

Each RequestHandlerWrapper takes a mapping of version to handler, where each version/handler item comes from the the routes which belong to that regex pattern.

So it looks like:

```javascript
{
    regex: [
        (RouteNode1, RequestHandlerWrapper({
            1: _HandlerForV1 // RequestHandler from RouteNode1
            2: _HandlerForV2 // RequestHandler from RouteNode2
        }),
        (RouteNode2, RequestHandlerWrapper({
            1: _HandlerForV1 // RequestHandler from RouteNode1
            2: _HandlerForV2 // RequestHandler from RouteNode2
        }),
    ],
    regex: [ /* ... */ ],
}
```

So the logic of creating this would look like

```python
route_handler_map = {}
for regex in route_map:
    # Create a dict that maps version to handler for each unique regex pattern
    handlers_map = dict([(route.version, route.handler) for route in route_map[regex]])
    for route in route_map[regex]:
        if regex not in route_handler_map:
            route_handler_map[regex] = []
        route_handler_map[regex].append((route, RequestHandlerWrapper(handlers_map)))
```

So we kind of loop through the route list in the `route_map` -- the first time being to create the RequestHandlerWrapper for each of the RouteNodes for that regex pattern, and again to assign each of those RouteNodes that RequestHandlerWrapper.